valblk: out-of-bounds in Writer.AddValue

https://mesolite.cluster.engflow.com/invocations/default/c85fe7d9-412e-4b8d-93f6-1d0f6db96038
```
panic: runtime error: slice bounds out of range [:6947] with capacity 6912
	panic: Release called on a BufferPool with in-use buffers
goroutine 115974 [running]:
github.com/cockroachdb/pebble/sstable/block.(*BufferPool).Release(0xc0180bb8c8)
	external/com_github_cockroachdb_pebble/sstable/block/buffer_pool.go:188 +0xc5
panic({0x6d68260?, 0xc0113bfdd0?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/pebble/sstable/block.(*Buffer).Append(0xc00bfb3ad0?, {0xc01039ea00?, 0x141a, 0x141a?})
	external/com_github_cockroachdb_pebble/sstable/block/compression.go:343 +0xf8
github.com/cockroachdb/pebble/sstable/valblk.(*Writer).AddValue(0xc00ffd3008?, {0xc01039ea00?, 0x141a, 0xc0118426c0?})
	external/com_github_cockroachdb_pebble/sstable/valblk/writer.go:80 +0xb6
github.com/cockroachdb/pebble/sstable.(*RawColumnWriter).AddWithForceObsolete(0xc00ffd3008, {{0xc0118426c0?, 0x39?, 0x60?}, 0x73e?}, {0xc01039ea00, 0x141a, 0x1500}, 0x0)
	external/com_github_cockroachdb_pebble/sstable/colblk_writer.go:357 +0x1ca
github.com/cockroachdb/pebble/internal/compact.(*Runner).writeKeysToTable(0xc01038b678, {0x8b13720, 0xc00ffd3008})
	external/com_github_cockroachdb_pebble/internal/compact/run.go:222 +0x7bc
github.com/cockroachdb/pebble/internal/compact.(*Runner).WriteTable(0xc01038b678, {0x2e, 0x2, {0x0, 0x0, {0x0, 0x0}, 0x0, {0x0, 0x0}, ...}}, ...)
	external/com_github_cockroachdb_pebble/internal/compact/run.go:151 +0x1ca
github.com/cockroachdb/pebble.(*DB).compactAndWrite(0xc007862488, 0x28, 0xc0180bb808, {0x0, 0x0, 0x0}, 0x7)
	external/com_github_cockroachdb_pebble/compaction.go:3034 +0x7b1
github.com/cockroachdb/pebble.(*DB).runCompaction(0xc007862488, 0x28, 0xc0180bb808)
	external/com_github_cockroachdb_pebble/compaction.go:2917 +0x725
github.com/cockroachdb/pebble.(*DB).compact1(0xc007862488, 0xc0180bb808, 0x0)
	external/com_github_cockroachdb_pebble/compaction.go:2350 +0x19e
github.com/cockroachdb/pebble.(*DB).compact.func1({0x8a9d478?, 0xc00a669b60?})
	external/com_github_cockroachdb_pebble/compaction.go:2264 +0xa7
runtime/pprof.Do({0x8a9d128?, 0xd57f000?}, {{0xc00cb3d3e0?, 0x1620200?, 0x8a48e40?}}, 0xc008390f88)
	GOROOT/src/runtime/pprof/runtime.go:51 +0x9d
github.com/cockroachdb/pebble.(*DB).compact(0xc007862488, 0xc0180bb808, 0x0)
	external/com_github_cockroachdb_pebble/compaction.go:2261 +0x7c
created by github.com/cockroachdb/pebble.(*DB).tryScheduleAutoCompaction in goroutine 88043
	external/com_github_cockroachdb_pebble/compaction.go:1926 +0x5c5
goroutine 1 [chan receive]:
```

Jira issue: PEBBLE-321

--- a/sstable/block/compression.go
+++ b/sstable/block/compression.go
@@ -338,7 +338,7 @@ func (b *Buffer) Append(v []byte) int {
 		for size < newLen {
 			size *= 2
 		}
-		b.h.b = slices.Grow(b.h.b, size-cap(b.h.b))
+		b.h.b = slices.Grow(b.h.b, size-len(b.h.b))
 	}
 	b.h.b = b.h.b[:newLen]
 	if n := copy(b.h.b[off:], v); n != len(v) {
@@ -411,10 +411,10 @@ func (h *BufHandle) Release() {
 		panic(errors.AssertionFailedf("pool has no maximum size"))
 	}
 	// Note we avoid releasing buffers that are larger than the configured
-	// maximum to the pool. This avoids holding on to occassional large buffers
-	// necesary for, for example, single large values.
+	// maximum to the pool. This avoids holding on to occasional large buffers
+	// necessary for, for example, singlular large values.
 	if h.b != nil && len(h.b) < h.pool.Max {
-		if invariants.Enabled {
+		if invariants.Sometimes(50) {
 			// Set the bytes to a random value. Cap the number of bytes being
 			// randomized to prevent test timeouts.
 			l := min(cap(h.b), 1000)
@@ -423,7 +423,7 @@ func (h *BufHandle) Release() {
 				h.b[j] = byte(rand.Uint32())
 			}
 		}
-		h.pool.pool.Put(h)
+		h.pool.Put(h)
 	}
 }
 
@@ -446,12 +446,29 @@ type bufferSyncPool struct {
 	pool    sync.Pool
 }
 
+// Put returns a buffer to the pool. While the buffer is in the pool, its pool
+// member is zeroed. This is used to validate invariants around double use of a
+// buffer.
+func (p *bufferSyncPool) Put(bh *BufHandle) {
+	if bh.pool != p {
+		panic(errors.AssertionFailedf("buffer has pool %v; trying to return it to pool %v", bh.pool, p))
+	}
+	bh.pool = nil
+	p.pool.Put(bh)
+}
+
 // Get retrieves a new buf from the pool, or allocates one of the configured
 // default size if the pool is empty.
 func (p *bufferSyncPool) Get() *BufHandle {
 	v := p.pool.Get()
 	if v != nil {
-		return v.(*BufHandle)
+		bh := v.(*BufHandle)
+		if bh.pool != nil {
+			panic(errors.AssertionFailedf("buffer has a pool; was it inserted into a pool twice?"))
+		}
+		// Set the pool so we know where to return the buffer to.
+		bh.pool = p
+		return bh
 	}
 	if invariants.Enabled && p.Default == 0 {
 		// Guard against accidentally forgetting to initialize a buffer sync pool.

--- a/sstable/block/compression_test.go
+++ b/sstable/block/compression_test.go
@@ -6,6 +6,7 @@ package block
 
 import (
 	"encoding/binary"
+	"fmt"
 	"math/rand/v2"
 	"testing"
 	"time"
@@ -91,4 +92,46 @@ func decompress(algo CompressionIndicator, b []byte) (*cache.Value, error) {
 		return nil, err
 	}
 	return decoded, nil
+}
+
+func TestBufferRandomized(t *testing.T) {
+	seed := uint64(time.Now().UnixNano())
+	t.Logf("seed %d", seed)
+	rng := rand.New(rand.NewPCG(0, seed))
+
+	var b Buffer
+	b.Init(SnappyCompression, ChecksumTypeCRC32c)
+	defer b.Release()
+	vbuf := make([]byte, 0, 1<<10) // 1 KiB
+
+	for i := 0; i < 25; i++ {
+		t.Run(fmt.Sprintf("iteration %d", i), func(t *testing.T) {
+			// Randomly release and reinitialize the buffer.
+			if rng.IntN(5) == 1 {
+				b.Release()
+				b.Init(SnappyCompression, ChecksumTypeCRC32c)
+			}
+
+			aggregateSizeOfKVs := rng.IntN(4<<20-(1<<10)) + 1<<10 // [1 KiB, 4 MiB)
+			size := 0
+			for b.Size() < aggregateSizeOfKVs {
+				vlen := rng.IntN(aggregateSizeOfKVs-b.Size()) + 1
+				if cap(vbuf) < vlen {
+					vbuf = make([]byte, vlen)
+				} else {
+					vbuf = vbuf[:vlen]
+				}
+				for i := range vbuf {
+					vbuf[i] = byte(rng.Uint32())
+				}
+				b.Append(vbuf)
+				size += vlen
+				require.Equal(t, size, b.Size())
+				s := b.Get()
+				require.Equal(t, vbuf, s[len(s)-len(vbuf):])
+			}
+			_, bh := b.CompressAndChecksum()
+			bh.Release()
+		})
+	}
 }

